Adds onChangeClientState callback prop

jasonblanchard opened the pull request:
This adds an optional `onUpdateTitle` callback that is invoked whenever the page title is updated. One use case for this would be to enable an accessibility library like this https://github.com/patrickfox/a11y_kit to hook into title changes and announce them to a screen reader, making page transitions more accessible (inspired by https://github.com/Robdel12/a11y-announcer).

It accomplishes this by calling `onUpdateTitle(document.title);` in the internal `updateTitle()` function in `Helmet.jsx`.

If folks like this approach, I'm happy to update the documentation and add callbacks to the other attributes that react-helmet can update (`base`, `meta`, etc), although I don't have use-cases in mind for those.

cwelch5 commented:
Thanks for the PR!  Here's the link to the CLA - https://cla.nfl.com/agreements/nfl/react-helmet (documentation in Helmet coming soon)

I definitely see the need.  I'm wondering if `PlainComponent.handleClientStateChangeCallback` couldn't be used to detect when there are DOM changes (such as title).  How we use in the unit tests is like this:
```
import {PlainComponent} from "../Helmet";

const oldHCSCC = PlainComponent.handleClientStateChangeCallback;
PlainComponent.handleClientStateChangeCallback = (state) => {
     // use state.title or any other state variable here to whatever use case
     return oldHCSCC(state);
};
```
I tried it in an app, just now, and realized we would need to export PlainComponent through index.jsx.  We can also put some more thought into making this as friendly and usable as possible.

But one plus with this, is having one callback for ANY DOM change instead of separate callbacks you have to manage.  Not having to pass callbacks to Helmet, keeps the API cleaner.

potench commented on src/Helmet.jsx:
```diff
@@ -41,6 +41,10 @@ const getTitleFromPropsList = (propsList) => {
     return innermostTitle || "";
 };
 
+const getonUpdateTitleFromPropsList = (propsList) => {
```
Should be camel-case: `getOnUpdateTitleFromPropsList`

potench commented on src/test/HelmetTest.jsx:
```diff
@@ -140,6 +140,32 @@ describe("Helmet", () => {
 
+        describe("onUpdateTitle", () => {
+            it("calls the function when the title udpates", () => {
+                const spy1 = sinon.spy();
```
Nice use of sinon for this. Can you change `spy1` to just `titleSpy`?

potench commented on src/test/HelmetTest.jsx:
```diff
@@ -140,6 +140,32 @@ describe("Helmet", () => {
+            });
+
+            it("calls uses the title template", () => {
+                const spy1 = sinon.spy();
```
Can you change `spy1` to `titleSpy` here?

potench commented on src/Helmet.jsx:
```diff
@@ -248,6 +255,7 @@ const reducePropsToState = (propsList) => {
 
     return {
         title: getTitleFromPropsList(propsList),
+        onUpdateTitle: getonUpdateTitleFromPropsList(propsList),
```
camel-case `getOnUpdateTitleFromPropsList`

jasonblanchard commented:
Thanks for the feedback! Glad to hear this is something that sounds useful.

Let me put some thought into how to possibly leverage `PlainComponent` here. My initial thought is that it feels weird to handle this callback as a static method on a component other than the `<Helmet>` component you are using in the codebase.

That said, I like the idea of using a single callback to handle state changes for all the properties that react-helmet can modify.

For now, I've made small changes to the in-line suggestions.

jasonblanchard commented:
Alright, I've updated this implementation to use a more general `onChangeClientState` callback that is invoked in `handleClientStateChange` and is passed the entire `newState` object. Usage would look like this:

```jsx
<Helmet
  title="Other Page"
  onChangeClientState={({title}) => announcePageTransition(title)}
/>
```

jasonblanchard commented on src/Helmet.jsx:
```diff
@@ -265,6 +272,8 @@ const handleClientStateChange = (newState) => {
     updateTags(TAG_NAMES.META, metaTags);
     updateTags(TAG_NAMES.BASE, baseTag);
 
+    onChangeClientState(newState);
```
Should this be more selective about which properties of `newState` are getting passed to the callback? Something like:

```javascript
onChangeClientState({title, baseTag, metaTags, linkTags, scriptTags});
```

cwelch5 replied:
I actually like how you did it.  No need to separate them out.  Whatever state coming in, should go unfiltered out to the user.

cwelch5 commented:
I had posted my own PR #72 right after I saw your initial PR, and now that I've been putting more thought into it, I'm hung up on one thing.  If I have 3 components that are nested, for example, and I only specify callbacks on the first and third Helmets.  When I change routes from 3 to 2, 3 will be unmounted, and on a client state change, because 2 doesn't have a callback, it will call the callback for 1.  That could have some adverse side effects (no pun intended :) )  I'm trying to brainstorm a way to alleviate this to make sure it always calls the callback for the deepest component you are currently on, and not call anything if there is no callback specified for that Helmet.

jasonblanchard commented:
I think I know what you mean, but just to clarify:

```jsx
class InnerComponent extends Component {
  render() {
    // This helmet does not have a callback.
    return <div><Helmet title="Inner Component" /></div>;
  }
}

class OuterComponent extends Component {
  render() {
    return (
      <div>
        <Helmet title="Outer Component" onChangeClientState={({title}) => console.log(title)} />
        <InnerComponent />
      </div>
    );
  }
}

class Wrapper extends Component {
  render() {
    return (
      <div>
        <Helmet title="Wrapper component" onChangeClientState={({title}) => console.log(title)} />
        <OuterComponent />
      </div>
    )
  }
}
```

When this tree is rendered via `<Wrapper>`, we **should not** see anything logged to the console because the inner-most component, `<InnerComponent>`, does not have a callback. As implemented, we **will** see `Wrapper component` and `Outer Component` logged to the console because their `handleClientStateChange` each fired `onChangeClientState`.

Does this replicate the problem you are describing?

jasonblanchard commented:
The above code actually logs out: `Wrapper component`, `Outer Component`, and `Inner Component`, the last one being called because it's using the deepest declared `onChangeClientState` prop. I think this is actually expected given how `getInnermostProperty` works inside `Helmet.jsx`, although I agree that it's probably not desirable behavior for this callback API.

Let me put some more thought into it, I think there's a way to handle this...

jasonblanchard commented:
@cwelch5 Just pushed up an implementation that *kind of* fixes the above scenario. There's a test case for it here https://github.com/nfl/react-helmet/pull/68/files#diff-9c0f68070060f3394d72fc17315fb2a4R169, but basically, for each `<Helmet>` in the render tree, it will only call the callback if the helmet has `onChangeClientState` defined.

It's not exactly the ideal case where only the deepest callback of all the helmets are called, but maybe that's something the calling code should handle. For my primary use case of announcing a title change, for example, this would be ok.

cwelch5 commented on src/Helmet.jsx:
```diff
@@ -29,6 +29,11 @@ const getInnermostProperty = (propsList, property) => {
+    const deepestProps = [...propsList].reverse()[0];
+    return deepestProps ? deepestProps[property] : null;
+};
+
```
This works perfect, but maybe we can rename it to make it more reusable.  `getInnermost` has the same prefix of `getInnermostProperty` which may confuse people thinking they do the same thing.  And having `callback` in the name suggests it can only be used for callbacks.  What do you think about something with property, to make it more general - `getStrictInnermostProperty`?  Ahh, naming is so hard.  Open to suggestions.

cwelch5 commented:
@jasonblanchard After seeing how we handled titleTemplate.  If we wanted to stay consistent, we should probably go with your previous commit where it always finds the deepest one regardless of if that is the Helmet you are currently seeing.  It's kind of a "set it and forget it" approach.  And I can see the ease of using it this way.  You can always overwrite titleTemplate or the callback downstream to change the functionality for a particular component.  But if you just set it once in the parent you never have to worry about it.  Sorry for the back and forth.  Do you agree?

jasonblanchard commented:
@cwelch5 No worries! Thanks for talking this through with me.

But yeah, I agree, I think that's more consistent with the API and I can see it being useful.

I've reverted the implementation and added a test case showing how this works.

mellowd3 commented:
Thanks for adding this! I am using it and relying on the document.title for Omniture pageview tracking data. However, it is logging out the same title two times onChangeClientState -- I'm only using this on views that switch content inside of them although the view is global for all. Otherwise I am not using this.
